evm/src/cpu/stack.rs

```diff
@@ -0,0 +1,387 @@
+    blockhash: None,        // TODO
+    coinbase: None,         // TODO
+    timestamp: None,        // TODO
+    number: None,           // TODO
+    difficulty: None,       // TODO
+    gaslimit: None,         // TODO
+    chainid: None,          // TODO
+    selfbalance: None,      // TODO
+    basefee: None,          // TODO
+    prover_input: None,     // TODO
+    pop: None,              // TODO
+    mload: None,            // TODO
+    mstore: None,           // TODO
+    mstore8: None,          // TODO
+    sload: None,            // TODO
+    sstore: None,           // TODO
+    jump: None,             // TODO
+    jumpi: None,            // TODO
+    pc: None,               // TODO
+    msize: None,            // TODO
+    gas: None,              // TODO
+    jumpdest: None,         // TODO
+    get_state_root: None,   // TODO
+    set_state_root: None,   // TODO
+    get_receipt_root: None, // TODO
+    set_receipt_root: None, // TODO
+    push: None,             // TODO
+    dup: None,              // TODO
+    swap: None,             // TODO
+    log0: None,             // TODO
+    log1: None,             // TODO
+    log2: None,             // TODO
+    log3: None,             // TODO
+    log4: None,             // TODO
+    create: None,           // TODO
+    call: None,             // TODO
+    callcode: None,         // TODO
+    return_: None,          // TODO
+    delegatecall: None,     // TODO
+    create2: None,          // TODO
+    get_context: None,      // TODO
+    set_context: None,      // TODO
+    consume_gas: None,      // TODO
+    exit_kernel: None,      // TODO
+    staticcall: None,       // TODO
+    mload_general: None,    // TODO
+    mstore_general: None,   // TODO
+    revert: None,           // TODO
+    selfdestruct: None,     // TODO
+    invalid: None,          // TODO
+};
+
+fn eval_packed_one<P: PackedField>(
+    lv: &CpuColumnsView<P>,
+    filter: P,
+    stack_behavior: StackBehavior,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let num_operands = stack_behavior.num_pops + (stack_behavior.pushes as usize);
+    assert!(num_operands <= NUM_GP_CHANNELS);
+
+    // Pops
+    for i in 0..stack_behavior.num_pops {
+        let channel = lv.mem_channels[i];
+
+        yield_constr.constraint(filter * (channel.used - P::ONES));
+        yield_constr.constraint(filter * (channel.is_read - P::ONES));
+
+        yield_constr.constraint(filter * (channel.addr_context - lv.context));
+        yield_constr.constraint(
+            filter * (channel.addr_segment - P::Scalar::from_canonical_u64(Segment::Stack as u64)),
+        );
+        // E.g. if `stack_len == 1` and `i == 0`, we want `add_virtual == 0`.
+        let addr_virtual = lv.stack_len - P::Scalar::from_canonical_usize(i + 1);
+        yield_constr.constraint(filter * (channel.addr_virtual - addr_virtual));
+    }
+
+    // Pushes
+    if stack_behavior.pushes {
+        let channel = lv.mem_channels[NUM_GP_CHANNELS - 1];
```

Comment:
So the order is (pops, non-stack ops, pushes)? Feels a tad unusual to me but it's fine if you prefer it.